File: src/main/java/com/cse3063f25grp1/answer/TemplateAnswerAgent.java
package com.cse3063f25grp1.answer;

import com.cse3063f25grp1.data.ChunkStore;
import com.cse3063f25grp1.model.Answer;
import com.cse3063f25grp1.model.Chunk;
import com.cse3063f25grp1.model.Hit;

import java.util.ArrayList;
import java.util.List;


public class TemplateAnswerAgent implements AnswerAgent {

    public TemplateAnswerAgent() {
    }

    @Override
    public Answer answer(List<String> queryTerms, List<Hit> topHits, ChunkStore chunkStore) {
        if (topHits == null || topHits.isEmpty()) {
            return new Answer("Üzgünüm, sorunuza cevap bulamadım.", new ArrayList<>());
        }

        // Get best chunk (first in reranked list)
        Hit bestHit = topHits.get(0);
        Chunk bestChunk = chunkStore.getChunk(bestHit.getDocId(), bestHit.getChunkId());
        if (bestChunk == null) {
            return new Answer("Üzgünüm, sorunuza ait detaylı metni bulamadım.", new ArrayList<>());
        }

        // Select best sentence from chunk
        String bestSentence = selectBestSentence(bestChunk.getText(), queryTerms);

        // Format answer
        String answerText = "Cevabınız: " + bestSentence;

        // Generate citations
        List<String> citations = new ArrayList<>();
        for (int i = 0; i < Math.min(3, topHits.size()); i++) {
            Hit hit = topHits.get(i);
            Chunk chunk = chunkStore.getChunk(hit.getDocId(), hit.getChunkId());
            if (chunk != null) {
                String citation = formatCitation(chunk);
                citations.add(citation);
            }
        }

        Answer answer = new Answer(answerText, citations);
        return answer;
    }

    /**
     * Select the best sentence from chunk text
     * Choose sentence with most query terms, or first sentence
     */
    private String selectBestSentence(String text, List<String> queryTerms) {
        if (text == null || text.isEmpty()) {
            return "Bilgi bulunamadı.";
        }

        // Split into sentences
        String[] sentences = text.split("[.!?]+");

        if (sentences.length == 0) {
            return text.substring(0, Math.min(200, text.length()));
        }

        String bestSentence = sentences[0].trim();
        int maxTermCount = 0;

        // Find sentence with most query terms
        for (String sentence : sentences) {
            sentence = sentence.trim();
            if (sentence.isEmpty())
                continue;

            int termCount = countQueryTerms(sentence, queryTerms);

            if (termCount > maxTermCount) {
                maxTermCount = termCount;
                bestSentence = sentence;
            }
        }

        return bestSentence;
    }

    /**
     * Count how many query terms appear in sentence
     */
    private int countQueryTerms(String sentence, List<String> queryTerms) {
        if (queryTerms == null)
            return 0;

        String lowerSentence = sentence.toLowerCase();
        int count = 0;

        for (String term : queryTerms) {
            if (lowerSentence.contains(term.toLowerCase())) {
                count++;
            }
        }

        return count;
    }

    /**
     * Format citation as: docId:sectionId:startOffset-endOffset
     */
    private String formatCitation(Chunk chunk) {
        return String.format("%s:%s:%d-%d",
                chunk.getDocId(),
                chunk.getSectionId(),
                chunk.getStartOffset(),
                chunk.getEndOffset());
    }
}

File: iteration1_py/src/answer/template_answer_agent.py
import re
from typing import List

from src.answer.answer_agent import AnswerAgent
from src.model.answer import Answer
from src.model.hit import Hit
from src.model.chunk import Chunk
from src.data.chunk_store import ChunkStore


class TemplateAnswerAgent(AnswerAgent):
    def __init__(self):
        pass
    
    def answer(self, query_terms: List[str], top_hits: List[Hit], chunk_store: ChunkStore) -> Answer:
        # If no hits → fallback
        if not top_hits:
            return Answer("Üzgünüm, sorunuza cevap bulamadım.", [])

        # Get best hit (first after reranking)
        best_hit = top_hits[0]
        best_chunk = chunk_store.get_chunk(best_hit.get_doc_id(), best_hit.get_chunk_id())

        if best_chunk is None:
            return Answer("Üzgünüm, sorunuza ait detaylı metni bulamadım.", [])

        # Select sentence
        best_sentence = self.__select_best_sentence(best_chunk.get_text(), query_terms)

        # Build source description
        doc_title = chunk_store.get_document_title(best_chunk.get_doc_id())

        if doc_title:
            source = f'Bu cevap "{doc_title}" başlıklı belgenin {best_chunk.get_section_id()} bölümünden alınmıştır. '
        else:
            source = f'Bu cevap {best_chunk.get_doc_id()} belgesinin {best_chunk.get_section_id()} bölümünden alınmıştır. '

        answer_text = source + "Cevabınız: " + best_sentence

        # Citations (first 3)
        citations: List[str] = []
        for hit in top_hits[:3]:
            chunk = chunk_store.get_chunk(hit.get_doc_id(), hit.get_chunk_id())
            if chunk:
                citations.append(self._format_citation(chunk))

        return Answer(answer_text, citations)

    # --------------------------
    # Private helpers
    # --------------------------

    def __select_best_sentence(self, text: str, query_terms: List[str]) -> str:
        if not text:
            return "Bilgi bulunamadı."

        sentences = [s.strip() for s in re.split(r"[.!?]+", text) if s.strip()]
        if not sentences:
            return text[:200]

        best_sentence = sentences[0]
        best_term_count = 0
        best_contains_all = False
        best_length = len(best_sentence)

        # Normalize query terms
        normalized_terms = [t.lower() for t in query_terms if t and t.strip()] if query_terms else []

        for sentence in sentences:
            term_count = self.__count_query_terms(sentence, normalized_terms)

            # Does this sentence contain all terms?    
            lower_sentence = sentence.lower()
            contains_all = all(term in lower_sentence for term in normalized_terms)

            length = len(sentence)

            # Main selection logic
            if contains_all:
                if (not best_contains_all or
                    term_count > best_term_count or
                    (term_count == best_term_count and length < best_length)):
                    best_contains_all = True
                    best_term_count = term_count
                    best_length = length
                    best_sentence = sentence
            elif not best_contains_all:
                if (term_count > best_term_count or
                    (term_count == best_term_count and length < best_length)):
                    best_term_count = term_count
                    best_length = length
                    best_sentence = sentence

        return best_sentence

    def __count_query_terms(self, sentence: str, query_terms: List[str]) -> int:
        if not query_terms:
            return 0

        lower_sentence = sentence.lower()
        return sum(1 for term in query_terms if term in lower_sentence)

    def _format_citation(self, chunk: Chunk) -> str:
        return f"{chunk.get_doc_id()}:{chunk.get_section_id()}:{chunk.get_start_offset()}-{chunk.get_end_offset()}"
